homeassistant/components/recorder/purge.py

```diff
@@ -101,3 +65,79 @@ def purge_old_data(instance, purge_days: int, repack: bool) -> bool:
     except SQLAlchemyError as err:
         _LOGGER.warning("Error purging history: %s", err)
     return True
+
+
+def _select_event_ids_to_purge(session: Session, purge_before: datetime) -> list:
+    """Return a list of event ids to purge."""
+    events = (
+        session.query(Events.event_id)
+        .filter(Events.time_fired < purge_before)
+        .limit(MAX_ROWS_TO_PURGE)
+        .all()
+    )
+    _LOGGER.debug("Selected %s event ids to remove", len(events))
+    return [event.event_id for event in events]
+
+
+def _select_state_ids_to_purge(
+    session: Session, purge_before: datetime, event_ids: list
+) -> list:
+    """Return a list of state ids to purge."""
+    if not event_ids:
+        return []
+    states = (
+        session.query(States.state_id)
+        .filter(States.last_updated < purge_before)
+        .filter(States.event_id.in_(event_ids))
+        .all()
+    )
+    _LOGGER.debug("Selected %s state ids to remove", len(states))
+    return [state.state_id for state in states]
+
+
+def _disconnect_states_about_to_be_purged(session: Session, state_ids: list) -> None:
```

Comment:
You'll probably need to pass `purge_before` here, too.

Reply:
Yup :)

Reply:
fixed